added test methods

OndrejSulc opened the pull request.

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("7");
```
Where is the cycle? Why should this test throw an exception?

quigamdev replied:
Here is no cycle. If this throws an exception then there is a bug in the detection of cycles.

exyi replied:
@OndrejSulc wrote:
> this specific test is not cyclic and should not throw error. I think you
copy pasted wrong one because there are two similar ones. This one you sent
me and other which has resource "dotvvm" dependent on resource "7" (makes it
cyclic). So you are looking for ResourceRepository_CyclicDependency_
Structure_2_Throws(). This is ResourceRepository_CyclicDependency_Structure_
1() and is not cyclic as you said correctly.

Well, you have the test inside the `Assert.ThrowsException<DotvvmResourceException>(() => { ... })` method that would fail the test if it did not throw. If you meant to test that it does not throw, you should remove the invocation of `Assert.ThrowsException`. The `_Throws` suffix in the name is ignored by testing framework, the names are just for humans.

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "DotVVM" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "DotVVM" } });
+
+                configuration.Resources.Register("DotVVM", new NullResource { Dependencies = new[] { "7" } }); ;//cyclic
```
I don't think it's a good idea to name test resources `DotVVM`.

exyi replied:
@OndrejSulc:
> 2) renamed to "dotvvm"

come on... What I meant it that the resource name should not be the same as another resource that already exists. If you mean to test something specific about the case when the resource already exists, please create another specific test for that, or at least add a short explanation what should happen.

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2", "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "5", "1" } });// cyclic
```
The resource `"1"` does not exist, so this test will probably throw even if I'd remove the cycle check. When you add the missing resource, this test is probably completely redundant with the test `ResourceRepository_CyclicDependency_Structure_2_Throws`

exyi replied:
@OndrejSulc:
> 3) resource "1" is "dotvvm". I didn't know how specified names of resources
have to be. Renamed to "dotvvm"

No, "1" should not refer to "dotvvm" by default. I have no idea what you mean...

exyi replied:
@OndrejSulc:
> What do you mean with redundant? It has different structure and Ladislav
Šesták didn't want just cycle or straight line.

I mean that it does not bring any value, these test would always together if we made any (reasonable) mistake in the code. The `ResourceRepository_CyclicDependency_Structure_2_Throws` is also not a simple path.

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("6");
+            });
```
How is this test different from `ResourceRepository_CyclicDependency_Structure_2_Throws`?

exyi replied:
@OndrejSulc:
> 4) Quite a lot I think. ResourceRepository_CyclicDependency_Structure_2_
Throws() has complicated structure but this one is simple tree.

It's not a tree since it has the cycle.

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");
+            });
```
How is this test different from `ResourceRepository_CyclicDependency_Structure_2_Throws`?

exyi replied:
@OndrejSulc:
> 5) I would expect question about how it is different compared to  ResourceRepository_CyclicDependency_Structure_4_Throws()

Ok, so how it's different from `ResourceRepository_CyclicDependency_Structure_4_Throws`? if you prefer.

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");
```
This is not cyclic, why does it throw?

exyi replied:
@OndrejSulc:
> 6)This method ResourceRepository_CyclicDependency_Structure_6() is not
specified as Throw().

The test framework ignores the name, what makes this test assert that the registration fails is the invocation of `Assert.ThrowsException` method.

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");
```
This is not cyclic, why does it throw?

exyi commented on src/DotVVM.Framework.Tests.Common/ResourceManagement/ResourceRepositoryTests.cs:
```diff
@@ -25,5 +25,199 @@ public void ResourceRepository_CyclicDependency_Throws()
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");
```
This is not cyclic, why does it throw?

OndrejSulc commented:
Hi,


1) this specific test is not cyclic and should not throw error. I think you 
copy pasted wrong one because there are two similar ones. This one you sent 
me and other which has resource "dotvvm" dependent on resource "7" (makes it
cyclic). So you are looking for ResourceRepository_CyclicDependency_
Structure_2_Throws(). This is ResourceRepository_CyclicDependency_Structure_
1() and is not cyclic as you said correctly.




2) renamed to "dotvvm"




3) resource "1" is "dotvvm". I didn't know how specified names of resources 
have to be. Renamed to "dotvvm"







4)





---------- Původní e-mail ----------
Od: Stanislav Lukeš <notifications@github.com>
Komu: riganti/dotvvm <dotvvm@noreply.github.com>
Datum: 23. 9. 2019 17:33:13
Předmět: Re: [riganti/dotvvm] added test methods (#763) 
"
@exyi requested changes on this pull request.

Apart from the in-code comments, could you please check that the error 
message is meaningful and check that it throws on a specific line, not in 
the entire test case?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327171203):

> +            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "6", "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4", "3" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("7");


1)Where is the cycle? Why should this test throw an exception?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327171798):

> +        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "6", "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4", "3" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "DotVVM" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "DotVVM" } });
+
+                configuration.Resources.Register("DotVVM", new NullResource { Dependencies = new[] { "7" } }); ;//cyclic


2)I don't think it's a good idea to name test resources DotVVM.
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327172977):

> +        public void ResourceRepository_CyclicDependency_Structure_3_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("8", new NullResource { Dependencies = new[] { "7", "6" } });
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5", "4" } });
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "3" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2", "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "5", "1" } });// cyclic


3)The resource "1" does not exist, so this test will probably throw even if 
I'd remove the cycle check. When you add the missing resource, this test is 
probably completely redundant with the test ResourceRepository_
CyclicDependency_Structure_2_Throws
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327179434):

> +        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_4_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "5" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "3" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "6" } });//cyclic
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("6");
+            });


How is this test different from ResourceRepository_CyclicDependency_
Structure_2_Throws?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327179604):

> +            });
+        }
+
+        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_5_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+                
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "3" } });//cyclic
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");
+            });


How is this test different from ResourceRepository_CyclicDependency_
Structure_2_Throws?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327179998):

> +            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "3", "1" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");


This is not cyclic, why does it throw?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327180245):

> +            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "7", "4", "3" } });
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "3", "1" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");


This is not cyclic, why does it throw?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327183532):

> +                configuration.Resources.Register("10", new NullResource { Dependencies = new[] { "9", "8", "7" } });
+
+                configuration.Resources.Register("9", new NullResource { Dependencies = new[] { "6", "5" } });
+                configuration.Resources.Register("8", new NullResource { Dependencies = new[] { "5" } });
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5","4" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "3", "2" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "2" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");


This is not cyclic, why does it throw?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
(https://github.com/riganti/dotvvm/pull/763?email_source=notifications&email_token=AHU5AZVLJ767WS3QWHSSKHTQLDOTHA5CNFSM4IZJWF32YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFSUZFY#pullrequestreview-291851415)
, or mute the thread
(https://github.com/notifications/unsubscribe-auth/AHU5AZTHOOF5DIDLX3ROZTLQLDOTHANCNFSM4IZJWF3Q)
. 
"

OndrejSulc commented:
"Hi,


1) this specific test is not cyclic and should not throw error. I think you 
copy pasted wrong one because there are two similar ones. This one you sent 
me and other which has resource "dotvvm" dependent on resource "7" (makes it
cyclic). So you are looking for ResourceRepository_CyclicDependency_
Structure_2_Throws(). This is ResourceRepository_CyclicDependency_Structure_
1() and is not cyclic as you said correctly.




2) renamed to "dotvvm"




3) resource "1" is "dotvvm". I didn't know how specified names of resources 
have to be. Renamed to "dotvvm"



"






Sorry.. Somehow it send it before I finished so

back to 3. : 

What do you mean with redundant? It has different structure and Ladislav 
Šesták didn't want just cycle or straight line.
""
 




4) Quite a lot I think. ResourceRepository_CyclicDependency_Structure_2_
Throws() has complicated structure but this one is simple tree.




5) I would expect question about how it is different compared 

to  ResourceRepository_CyclicDependency_Structure_4_Throws()
"



"
Same as 4.




6)This method ResourceRepository_CyclicDependency_Structure_6() is not 
specified as Throw().

 

7) same as 6

8) same as 6




We can make call in the evening or tomorow if you would like to.

Regards Ondřej Šulc






---------- Původní e-mail ----------
Od: Stanislav Lukeš <notifications@github.com>
Komu: riganti/dotvvm <dotvvm@noreply.github.com>
Datum: 23. 9. 2019 17:33:13
Předmět: Re: [riganti/dotvvm] added test methods (#763) 
"
@exyi requested changes on this pull request.

Apart from the in-code comments, could you please check that the error 
message is meaningful and check that it throws on a specific line, not in 
the entire test case?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327171203):

> +            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "6", "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4", "3" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("7");


1)Where is the cycle? Why should this test throw an exception?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327171798):

> +        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "6", "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4", "3" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "DotVVM" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "DotVVM" } });
+
+                configuration.Resources.Register("DotVVM", new NullResource { Dependencies = new[] { "7" } }); ;//cyclic


2)I don't think it's a good idea to name test resources DotVVM.
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327172977):

> +        public void ResourceRepository_CyclicDependency_Structure_3_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("8", new NullResource { Dependencies = new[] { "7", "6" } });
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5", "4" } });
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "3" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2", "1" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "5", "1" } });// cyclic


3)The resource "1" does not exist, so this test will probably throw even if 
I'd remove the cycle check. When you add the missing resource, this test is 
probably completely redundant with the test ResourceRepository_
CyclicDependency_Structure_2_Throws
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327179434):

> +        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_4_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "5" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "4" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "3" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "6" } });//cyclic
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("6");
+            });


4)How is this test different from ResourceRepository_CyclicDependency_
Structure_2_Throws?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327179604):

> +            });
+        }
+
+        [TestMethod]
+        public void ResourceRepository_CyclicDependency_Structure_5_Throws()
+        {
+            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+                
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "3" } });//cyclic
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");
+            });


5)How is this test different from ResourceRepository_CyclicDependency_
Structure_2_Throws?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327179998):

> +            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "4", "3" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "3", "1" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");


6)This is not cyclic, why does it throw?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327180245):

> +            Assert.ThrowsException<DotvvmResourceException>(() => {
+                var configuration = DotvvmConfiguration.CreateDefault();
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "7", "4", "3" } });
+
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "3", "1" } });
+
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "1" } });
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "2" } });
+
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");


7)This is not cyclic, why does it throw?
----------------------------------------------------------------------------

In src/DotVVM.Framework.Tests.Common/ResourceManagement/
ResourceRepositoryTests.cs
(https://github.com/riganti/dotvvm/pull/763#discussion_r327183532):

> +                configuration.Resources.Register("10", new NullResource { Dependencies = new[] { "9", "8", "7" } });
+
+                configuration.Resources.Register("9", new NullResource { Dependencies = new[] { "6", "5" } });
+                configuration.Resources.Register("8", new NullResource { Dependencies = new[] { "5" } });
+                configuration.Resources.Register("7", new NullResource { Dependencies = new[] { "5","4" } });
+
+                configuration.Resources.Register("6", new NullResource { Dependencies = new[] { "3", "2" } });
+                configuration.Resources.Register("5", new NullResource { Dependencies = new[] { "2" } });
+                configuration.Resources.Register("4", new NullResource { Dependencies = new[] { "2", "1" } });
+
+                configuration.Resources.Register("3", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("2", new NullResource { Dependencies = new[] { "dotvvm" } });
+                configuration.Resources.Register("1", new NullResource { Dependencies = new[] { "dotvvm" } });
+
+                var manager = configuration.ServiceProvider.GetRequiredService<ResourceManager>();
+                manager.AddRequiredResource("3");


8)This is not cyclic, why does it throw?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
(https://github.com/riganti/dotvvm/pull/763?email_source=notifications&email_token=AHU5AZVLJ767WS3QWHSSKHTQLDOTHA5CNFSM4IZJWF32YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFSUZFY#pullrequestreview-291851415)
, or mute the thread
(https://github.com/notifications/unsubscribe-auth/AHU5AZTHOOF5DIDLX3ROZTLQLDOTHANCNFSM4IZJWF3Q)
. 
"

ENgateman commented:
@OndrejSulc 

Why do the tests need to refer to the `dotvvm` resource?

The redundancy of some of these tests would maybe be a little more obvious if the tests had better names. Some of them don't even test for cycles, so why do they have "CyclicDependency" in their name?

"Structure_n" is somewhat useless too. I doubt that even you will remember what was the difference between "Structure_1" and "Structure_8" in a few weeks. Please try to describe them (e.g.: "Path", "Tree", "Cycle", "MultipleCycles", etc).
